File: app/src/main/java/com/gkgio/translate/helpers/errorHandilng/ErrorHandler.java
package com.gkgio.translate.helpers.errorHandilng;


import com.gkgio.translate.BuildConfig;
import com.gkgio.translate.di.scope.ActivityScope;

import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.util.concurrent.TimeoutException;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.functions.Function;
import retrofit2.HttpException;


@ActivityScope
public class ErrorHandler implements ObservableTransformer<Object, Object> {

    private final ErrorConverter errorConverter;
    private final ApiErrors errors;

    @Inject
    ErrorHandler(ErrorConverter errorConverter, ApiErrors errors) {
        this.errorConverter = errorConverter;
        this.errors = errors;
    }

    @Override
    public ObservableSource<Object> apply(Observable<Object> single) {
        return single.onErrorResumeNext(new Function<Throwable, Observable<?>>() {
            @Override
            public Observable<?> apply(Throwable e) {
                Throwable result;

                if (e instanceof HttpException) {

                    final HttpException he = (HttpException) e;

                    ErrorDTO error = null;
                    try {
                        error = errorConverter.getError(he);
                    } catch (Exception e1) {
                        // ignore
                    }

                    final int httpCode = he.code();
                    final int code = (error != null) ? error.code : -1;

                    result = new ServerException(httpCode, error.message, code);

                } else if (e instanceof TimeoutException) {
                    result = new ConnectionException(errors.getMessage(ApiErrors.SERVER_CONNECT_TIMEOUT), e);
                } else if (isConnectionError(e)) {
                    result = new ConnectionException(errors.getMessage(ApiErrors.NO_CONNECTION), e);
                } else {
                    result = new UnexpectedError(errors.getMessage(ApiErrors.UNEXPECTED_ERROR), e);
                }

                if (BuildConfig.DEBUG) {
                    result.printStackTrace();
                }

                return Observable.error(result);
            }
        });
    }

    @SuppressWarnings("unchecked")
    public <T> ObservableTransformer<T, T> cast() {
        return (ObservableTransformer<T, T>) this;
    }


    private boolean isConnectionError(Throwable e) {
        return (e instanceof TimeoutException) ||
                (e instanceof UnknownHostException) ||
                (e instanceof SocketTimeoutException) ||
                (e instanceof java.net.ConnectException);
    }
}

File: app/src/main/java/com/gkgio/translate/helpers/errorHandling/ErrorHandler.kt
package com.gkgio.translate.helpers.errorHandling


import com.gkgio.translate.BuildConfig
import com.gkgio.translate.di.scope.ActivityScope

import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.concurrent.TimeoutException

import javax.inject.Inject

import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.ObservableTransformer
import io.reactivex.functions.Function
import retrofit2.HttpException


@ActivityScope
class ErrorHandler @Inject
internal constructor(private val errorConverter: ErrorConverter,
                     private val errors: ApiErrors) : ObservableTransformer<Any, Any> {

  override fun apply(single: Observable<Any>): ObservableSource<Any> {
    return single.onErrorResumeNext(Function<Throwable, Observable<*>> { e ->
      val result: Throwable

      if (e is HttpException) {

        var error: ErrorDTO? = null
        try {
          error = errorConverter.getError(e)
        } catch (e1: Exception) {
          // ignore
        }

        val httpCode = e.code()
        val code = error?.code ?: -1

        result = ServerException(httpCode, error?.message ?: "", code)

      } else if (e is TimeoutException) {
        result = ConnectionException(errors.getMessage(ApiErrors.SERVER_CONNECT_TIMEOUT), e)
      } else if (isConnectionError(e)) {
        result = ConnectionException(errors.getMessage(ApiErrors.NO_CONNECTION), e)
      } else {
        result = UnexpectedError(errors.getMessage(ApiErrors.UNEXPECTED_ERROR), e)
      }

      if (BuildConfig.DEBUG) {
        result.printStackTrace()
      }

      Observable.error<Any>(result)
    })
  }

  @SuppressWarnings("unchecked")
  fun <T> cast(): ObservableTransformer<T, T> {
    return this as ObservableTransformer<T, T>
  }


  private fun isConnectionError(e: Throwable): Boolean {
    return e is TimeoutException ||
        e is UnknownHostException ||
        e is SocketTimeoutException ||
        e is java.net.ConnectException
  }
}
